Post 엔티티, Spring Rest Docs 기본 설정, Post 메인페이지(최신순으로 정렬기능)

102092 opened the pull request:
Resolved #7 , Resolved #11

102092 committed:
[#7] Chore : prod actuator 설정 추가

102092 committed:
[#7] Feat : Main Page Domain 생성

- Account User 역할, syntax 문제로 변경
- Like : 좋아요, 여러 유저는 여러개의 좋아요를 만들 수 있음. Relation 테이블 필요함

102092 committed:
[#7] Chore : ddl-auto 옵션 수정

- create -> create-drop

102092 committed:
[#7] Refactor : Entity 연관관계 제거

- 빠른 CRUD 기능 구현을 위해, 우선 연관관계 제거
- 기능 구현 후 추후 매핑 예정

102092 committed:
[#11] Feat : JackSon 설정 추가

- 기본 DateTimeFormatting 설정

102092 committed:
[#11] Feat : Main 페이지 조회를 위한 Post Model 작성

- MVC 구조를 위한 Service, Controller 작성
- Application Runner를 상속한 데이터 추가

102092 committed:
[#11] Feat : BaseTimeEntity 생성

- 생성, 시간을 관리할 추상 객체
- 현재는 Post만 상속하고 있음.

102092 committed:
[#11] Feat : ApiResult, Error 객체 생성

- Api 응답에 사용할 객체, Error 에 사용할 기본 객체를 생성함
- 추후 enum등을 이용해서 Error code를 핸들링할 생각

102092 committed:
[#11] Feat : RestDocs 적용

- Post FindAll() 메서드 에 대한 테스트 작성
- RestDocs 적용
- 추후 Docs의 template을 보기 좋게 수정할 예정

102092 committed:
[#11] Refactor : ApiError 접근제어자 수정

102092 committed:
[#11] Feat : 기본 Error Handle 설정

- Entity NotFound에 대한 에러 처리
- 추후, 인증등의 에러 핸들링 추가 예정

102092 committed:
[#11] Feat : Post Id로 조회 하는 기능 추가

- 조회 중 오류 발생 시(해당하는 Post를 못찾았을 경우), NotFoundException 발생

102092 committed:
[#11] Refactor : CommandLiner 수정

- DB와 연결되는 회수를 1번으로 수정

102092 committed:
[#11] Feat : 메인페이지 최신순 정렬 기능

- 기본적으로 생성시간 내림차순으로 진행됨
- 이를 위해 Post Entity에서 @CreationTimeStamp를 주석 처리함

hanurii commented on src/main/java/com/codesquad/rare/common/BaseTimeEntity.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
```
@MappedSuperclass 랑 @EntityListeners 가 무엇을 뜻하는 지 잘 모르겠습니다. 찾아보고 주석 달아보겠습니다.

102092 replied:
저도 정확히는 모르지만, 전자는 상속을 위한 Entity에, 후자는... 저도 찾아봐야겠네용..ㅠㅠ

hanurii commented on src/main/java/com/codesquad/rare/common/BaseTimeEntity.java:
```diff
@@ -0,0 +1,18 @@
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseTimeEntity {
+
+  @CreatedDate
```
@CreatedDate 이것도 무엇을 의미하는지 잘 모르겠습니다. 찾아보고 주석 달아볼게요.

hanurii commented on src/main/java/com/codesquad/rare/common/api/ApiError.java:
```diff
@@ -0,0 +1,23 @@
+  private final int status;
+
+  public ApiError(Exception exception, HttpStatus status) {
+    this(exception.getMessage(), status);
```
이렇게 작성하면 다음과 같은 역할을 하는 건가요?

```
public ApiError(Exception exception, HttpStatus status) {
    this.message = exception.getMessage();
    this.status = status;
}
```

102092 replied:
`this(...)` 를 통해서 해당 객체 내에 있는 다른 생성자를 이용하도록 하는 방법입니다

hanurii replied:
오 하나 알고갑니다!

hanurii commented on src/main/java/com/codesquad/rare/common/api/ApiError.java:
```diff
@@ -0,0 +1,23 @@
+
+  public ApiError(String message, HttpStatus status) {
+    this.message = message;
+    this.status = status.value();
```
인스턴스 생성할 때 다음과 같이 쓰이고 status.value() == 200 이라고 생각하면 될까요? 안써본 형태라 궁금하네요. 찾아보고 주석 달아보겠습니다.
```
ApiError apiError = new ApiError("SUCCESS", 200); 
```

102092 replied:
`ApiError` 는 `GlobalExceptionHandler` 에서 사용하는 클래스에요
위와 같은 방식으로 만들어진다기 보다는 저희가 Servie 혹은 Controller 클래스 `() -> new SomeException` 처럼 어떠한 오류를 발생 시키면, 그걸 `GlobalExceptionHandler` 에서 잡아서, ApiError 자동으로 만들어주는 형식으로 동작하고 있습니다.

괜찮으시면 만나서, 어떻게 동작하는지 같이 살펴봐도 좋겠네요 !

hanurii replied:
좋습니다 아직 잘 이해는 안가지만 우선 눈으로라도 익혀둘게요

hanurii commented on src/main/java/com/codesquad/rare/common/api/ApiResult.java:
```diff
@@ -0,0 +1,34 @@
+    this.error = error;
+  }
+
+  public static <T> ApiResult<T> OK(T response) {
+    return new ApiResult<>(true, response, null);
+  }
```
한님 코드를 보면 제네릭을 잘 활용하시는 것 같습니다. 제가 잘 몰라서 무슨 뜻인지 헷갈리는데 제가 공부해보고 주석 남겨보겠습니다.

hanurii commented on src/main/java/com/codesquad/rare/common/api/ApiResult.java:
```diff
@@ -0,0 +1,34 @@
+
+@Getter
+@ToString
+public class ApiResult<T> {
```
한님 코드에서 이런 형태의 ApiResult, ApiError 클래스를 잘 만들어서 활용하시던데 어떤 책에 나와있어서 차용해서 쓰시는 건가요? 저도 배워보고 싶네요!

hanurii commented on src/main/java/com/codesquad/rare/config/WebConfig.java:
```diff
@@ -0,0 +1,29 @@
+@Configuration
+public class WebConfig {
+
+  private static final String DATE_FORMAT = "yyyy-MM-dd";
+  private static final String DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss";
+  private static final String TIME_ZONE = "Asia/Seoul";
```
이렇게 설정해두면 프로그램 구동 시 시간이 설정되는 건가요. 궁금하네요!

hanurii commented on src/main/java/com/codesquad/rare/domain/Account.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Id
+  @Column(name = "user_id")
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
```
@GeneratedValue 에 strategy = GenerationType.IDENTITY 속성이 무슨 역할을 하는지 궁금하네요! 찾아보고 주석을 달아볼게요

hanurii commented on src/main/java/com/codesquad/rare/domain/Account.java:
```diff
@@ -0,0 +1,32 @@
+
+@Getter
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@NoArgsConstructor
```
@AllArgs ... @NoArgs ... @Require ... 3개의 역할과 사용법이 헷갈리는데 공부해보고 주석 남겨보겠습니다.

hanurii commented on src/main/java/com/codesquad/rare/domain/post/Post.java:
```diff
@@ -0,0 +1,58 @@
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@NoArgsConstructor
+@JsonPropertyOrder({
+    "id",
+    "title",
+    "content",
+    "thumbnail",
+    "author",
+    "views",
+    "likes",
+    "tags",
+    "createdTimeAt"
+})
```
오 신기한 애노테이션 사용법이네요 재밌습니다! 찾아서 주석에 달아볼게요!

102092 replied:
좀 깔끔한 json 형태를 찾다보니..
사실 Post DTO를 만들면, 없어질 부분일 것 같아용

hanurii commented on src/main/java/com/codesquad/rare/domain/post/PostDataRunner.java:
```diff
@@ -0,0 +1,55 @@
+
+@Slf4j
+@Component
+public class PostDataRunner implements ApplicationRunner {
```
ApplicationRunner 는 처음보는데요. 역할이 어떤지 궁금하네요.

102092 replied:
Application이 실행될때 같이 실행되는 친구입니다.
실행될 때, DB에 mock data를 넣어주고 있어요!

hanurii commented on src/main/java/com/codesquad/rare/domain/post/PostService.java:
```diff
@@ -0,0 +1,31 @@
+    this.postRepository = postRepository;
+  }
+
+  @Transactional(readOnly = true)
+  public List<Post> findAll() {
+    return postRepository.findAllByOrderByCreatedTimeAtDesc();
+  }
+
+  @Transactional(readOnly = true)
+  public Post findById(Long postId) {
+    return postRepository.findById(postId).orElseThrow(() -> new NotFoundException(Post.class, postId));
+  }
+
+  @Transactional
+  public Post save(Post post) {
+    return postRepository.save(post);
+  }
+}
```
김영한님 강의보고 생각난 건데요.

클래스 위에 @Transactional(readOnly = true) 를 써놓고
조회하는 메서드는 @트랜잭셔널 붙이지않고 값의 변동이 생기는 메서드(예:save()) 위에만 @Transactional 붙이는 형태 어떤신가요??

102092 replied:
좋습니다! 그렇게 하면 코드가 좀더 깔끔해질듯 해요!

hanurii commented on src/test/java/com/codesquad/rare/domain/post/PostControllerTest.java:
```diff
@@ -0,0 +1,133 @@
+
+  @DisplayName("메인 페이지 조회")
+  @Test
+  void 메인_페이지_조회() throws Exception {
```
메서드명이 한글일 경우 저는 snippets 이 생성되지 않던데 한님은 되셨으니까 제 intellij 한글 설정 관련 문제가 있는 것 같네요 ㅠㅠ

102092 replied:
ㅎㅎㅎ 저도 하밀이 작성하신 문서에서 확인하고 영어로 바꿨습니다!! 
감사합니다

hanurii replied:
감사합니다!

102092 committed:
[#11] Feat : api-guide.adoc 포멧 수정

- table content 추가
- 필요한 snippets만 추가함

hanurii reviewed (commented):
리뷰 테스트1

102092 committed:
[#11] Refactor : @Transactional 수정

- 코드 중복을 막기 위해, 클래스 위에 @Transactional(...) 작성
- readonly가 필요하지 않는 곳에는 @Transactional 명시

102092 committed:
[#11] Refactor : 사용하지 않는 메서드 삭제, ApiError 생성자의 접근제어자 수정

- ExceptionHandler에서 사용하지 않을 newResponse 메서드 삭제
- ApiError 생성자의 접근제어자 수정

102092 committed:
[#11] Refactor : 사용하지 않는 변수 삭제

- days 변수 삭제
- 코드 정렬
